src/Agent.Worker/Release/Artifacts/BuildArtifact.cs

```diff
@@ -181,26 +184,35 @@ private bool Match(ServerBuildArtifact buildArtifact, ArtifactDefinition artifac
                 var fileShareArtifact = new FileShareArtifact();
                 await fileShareArtifact.DownloadArtifactAsync(executionContext, HostContext, artifactDefinition, fileShare, downloadFolderPath);
             }
-            else if (string.Equals(buildArtifact.Resource.Type, WellKnownArtifactResourceTypes.Container, StringComparison.OrdinalIgnoreCase))
+            else if (buildArtifactDetails != null &
+                     string.Equals(buildArtifact.Resource.Type, WellKnownArtifactResourceTypes.Container, StringComparison.OrdinalIgnoreCase))
             {
-                executionContext.Output("Artifact Type: ServerDrop");
-
-                // TODO:Get VssBinFetchclient and get away from zipstream downloader
-                Stream contentStream;
+                string[] parts = buildArtifact.Resource.Data.Split(new[] { '/' }, 3);
 
-                if (definitionType == DefinitionType.Xaml)
-                {
-                    contentStream = await xamlBuildClient.GetArtifactContentZipAsync(buildArtifactDetails.Project, buildId, buildArtifact.Name);
-                }
-                else
+                int containerId;
+                string rootLocation = parts[2];
+                if (!int.TryParse(parts[1], out containerId))
                 {
-                    contentStream = await buildClient.GetArtifactContentZipAsync(buildArtifactDetails.Project, buildId, buildArtifact.Name);
+                    throw new ArtifactDownloadException(string.Format(CultureInfo.CurrentCulture, "Resources.NotContainerType", buildArtifact.Name));
                 }
 
+                using (var engine = new ContainerFetchEngine(
+                    executionContext,
+                    buildArtifactDetails.TfsUrl,
+                    buildArtifactDetails.Project,
+                    containerId,
+                    rootLocation,
+                    Path.Combine(localFolderPath, rootLocation),
+                    parallelDownloadLimit: 4,
+                    verbose: false,
+                    accessToken: buildArtifactDetails.AccessToken))
+                {
+                    var fetchTask = engine.FetchAsync();
 
-                var zipStreamDownloader = HostContext.GetService<IZipStreamDownloader>();
-                string artifactRootFolder = StringUtil.Format("/{0}", buildArtifact.Name);
-                await zipStreamDownloader.DownloadFromStream(contentStream, artifactRootFolder, buildArtifactDetails.RelativePath, downloadFolderPath);
+                    while (!fetchTask.Wait(TimeSpan.FromMinutes(20)))
+                    {
+                    }
```

Comment:
actually all this code is copy/pasted for ContainerFetchEngine, I couldn't get to refactoring\cleaning it since it required for RC1. Wait isn't needed here we can just do await engine.FetchAsync()